File: testing/PIMS.Tests.Automation/StepDefinitions/NotesSteps.cs

using OpenQA.Selenium;
using PIMS.Tests.Automation.Classes;
using PIMS.Tests.Automation.Data;
using PIMS.Tests.Automation.PageObjects;
using System.Data;

namespace PIMS.Tests.Automation.StepDefinitions
{
    [Binding]
    public class NotesSteps
    {
        private readonly Notes notes;
        private readonly SharedPagination sharedPagination;
        private readonly GenericSteps genericSteps;

        private List<string> notesData;
        private int notesCount;

        public NotesSteps(IWebDriver driver)
        {
            notes = new Notes(driver);
            sharedPagination = new SharedPagination(driver);
            genericSteps = new GenericSteps(driver);
            notesData = new List<string>();
            notesCount = 0;
        }

        [StepDefinition(@"I create a new Note on the Notes Tab from row number (.*)")]
        public void CreateNotesTab(int rowNumber)
        {
            /* TEST COVERAGE: PSP-5332, PSP-5505 */

            //Navigate to the Notes Tab
            notes.NavigateNotesTab();
            notes.VerifyNotesTabListView();

            //Create a new note
            PopulateNotes(rowNumber);
            notes.CreateNotesTabButton();
            notes.VerifyNotesAddNew();
            notes.AddNewNoteDetails(notesData[0]);

            //Cancel new note
            notes.CancelNote();

            //Create a new note
            for (var i = 0; i < notesData.Count; i++)
            {
                notes.CreateNotesTabButton();
                notes.AddNewNoteDetails(notesData[i]);
                notes.SaveNote();
            }
        }

        [StepDefinition(@"I edit a Note on the Notes Tab from row number (.*)")]
        public void EditNotesTab(int rowNumber)
        {
            /* TEST COVERAGE: PSP-4020, PSP-5506, PSP-5507 */

            //Navigate to the Notes Tab
            notes.NavigateNotesTab();

            //Edit note
            PopulateNotes(rowNumber);
            notes.ViewSecondNoteDetails();
            notes.VerifyNotesEditForm();
            notes.EditNote(notesData[0]);

            //Cancel note's update
            notes.CancelNote();

            //Edit note
            notes.ViewSecondNoteDetails();
            notes.EditNote(notesData[0]);

            //Save changes
            notes.SaveNote();

            //Verify Pagination on the list view
            sharedPagination.ChoosePaginationOption(5);
            Assert.Equal(5, notes.NotesTabCount());

            sharedPagination.ChoosePaginationOption(10);
            Assert.Equal(10, notes.NotesTabCount());

            //Delete Note
            notesCount = notes.NotesTabCount();
            notes.DeleteLastSecondNote();
        }

        private void PopulateNotes(int rowNumber)
        {
            DataTable notesSheet = ExcelDataContext.GetInstance().Sheets["Notes"]!;
            ExcelDataContext.PopulateInCollection(notesSheet);

            notesData = genericSteps.PopulateLists(ExcelDataContext.ReadData(rowNumber, "Notes"));
        }
    }
}

File: testing/playwright-testing/features/notes/notes.steps.js
const { Given, When } = require("@cucumber/cucumber");
const { expect } = require("@playwright/test");
const { splitStringToArray } = require("../../support/common.js");
const notesData = require("../../data/Notes.json");

let notesList = [];

Given(
  "I create a new Note on the Notes Tab with row number {int}",
  async function (rowNbr) {
    notesList = splitStringToArray(notesData[rowNbr].Notes);

    //Navigate to the Notes Tab
    await this.notes.navigateNotesTab();
    await this.notes.verifyNotesTabListView();

    //Create a new note
    await this.notes.createNotesTabButton();
    await this.notes.verifyNotesAddNew();
    await this.notes.addNewNoteDetails(notesList[0]);

    //Cancel new note
    await this.notes.cancelNote();

    //Create a new note
    for (const note of notesList) {
      await this.notes.createNotesTabButton();
      await this.notes.addNewNoteDetails(note);
      await this.notes.saveNote();
    }
  }
);

When(
  "I edit a Note on the Notes Tab with row number {int}",
  async function (rowNbr) {
    notesList = splitStringToArray(notesData[rowNbr].Notes);

    //Navigate to the Notes Tab
    await this.notes.navigateNotesTab();

    //Edit note
    await this.notes.viewSecondNoteDetails();
    await this.notes.editNoteButton();
    await this.notes.verifyNotesEditForm();
    await this.notes.editNote(notesList[0]);

    //Cancel note's update
    await this.notes.cancelNote();

    //Edit note
    await this.notes.viewSecondNoteDetails();
    await this.notes.editNoteButton();
    await this.notes.editNote(notesList[0]);

    //Save changes
    await this.notes.saveNote();

    //Verify Pagination on the list view
    await this.sharedPagination.choosePaginationOption(5);
    const notesCount5 = await this.notes.notesTabCount();
    expect(notesCount5).toBe(5);

    await this.sharedPagination.choosePaginationOption(10);
    const notesCount10 = await this.notes.notesTabCount();
    expect(notesCount10).toBe(10);

    //Delete Note
    await this.notes.deleteLastSecondNote();
  }
);
